Type test for ref struct is incorrect

If `ref struct G<T>` defined, `this` type test inside the struct for any resolved generic returns false (and compiler not even add test in IL).
It also shows warning CS0184 and does not allow to use it in pattern math type test.

**Version Used**:  C# 10

**Steps to Reproduce**:

Compile and run next code:

Use next code:

```csharp
using System;

new G<int>().Test();
new G<object>().Test();

ref struct G<T>
{
    public void Test()
    {
        if (this is G<int>)
        {
            Console.WriteLine("int");
        }
        else if (this is G<object>)
        {
            Console.WriteLine("object");
        }
        else
        {
            Console.WriteLine("unknown");
            Console.WriteLine(typeof(T));
        }
    }

    /*public void TestPattern()
    {
        var genericTypePattern = this switch
        {
            G<int> => "int",
            G<object> => "object",
            _ => "unknown"
        };

        Console.WriteLine(genericTypePattern);
    }*/
}
```
[On SharpLib.IO](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoavGAdwAIBxAHgEs8MA+ACgEoAdABUYAZwwCA3LQYtWEYACsYYHgJHjJ/GTSqwAZowm4184d2oBvaozuMiAZgcpGoiQNv2bVe38bsRrwYABbsYgERbJw8/F7+jD4JCURkAJy8AEQxmTrx/gC++X4wADZiMAFBoeGR8ooqatxxvslJyX6pGZkNqhi5uh1FrQllFcXeEynpWfgA1ngQ9HgDU/5dwQCeAA4wEAa8wvx5I4X5w/kA9ABUTi5uWgAKAIYYGDBQeJ6n7QkAbs8oIwAOYwOhQdhgYQ7GAvN4fPCMAC8jBqETE9HYGDAITWvw68hi3GRxOyXEyaDWfjYvSaJMYPWUfQpVPsAH16Zl5otlpk1gVdGsNqDwZDobs4e9Pic/AVrpdqAUgA==)


**Expected Behavior**:
Program should output:
```
int
object
```

No CS0184 warning should not be shown, if uncomment, TestPattern method should compile without errors.

**Actual Behavior**:
Program outputs:
```
unknown
System.Int32
unknown
System.Object
```

CS0184 warning shown, TestPattern method can not be compiled without changes

--- a/docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md
+++ b/docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md
@@ -1,5 +1,22 @@
 # This document lists known breaking changes in Roslyn after .NET 6 all the way to .NET 7.
 
+## Type tests for `ref` structs ar not supported.
+
+***Introduced in Visual Studio 2022 version 17.4***
+
+When a `ref` struct type is used in an 'is' or 'as' operator, in some scenarios compiler was previously reporting
+an erroneous warning about the type test always failing at runtime, omitting the actual type check, and leading to
+an incorrect behavior. When an incorrect behavior at execution time was possible, compiler will now produce an error instead.
+
+```csharp
+ref struct G<T>
+{
+    public void Test()
+    {
+        if (this is G<int>) // Will now produce an error, used to be treated as always `false`.
+        {
+```
+
 ## Unused results from ref local are dereferences.
 
 ***Introduced in Visual Studio 2022 version 17.4***

--- a/docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md
+++ b/docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md
@@ -6,7 +6,7 @@
 
 When a `ref` struct type is used in an 'is' or 'as' operator, in some scenarios compiler was previously reporting
 an erroneous warning about the type test always failing at runtime, omitting the actual type check, and leading to
-an incorrect behavior. When an incorrect behavior at execution time was possible, compiler will now produce an error instead.
+incorrect behavior. When incorrect behavior at execution time was possible, compiler will now produce an error instead.
 
 ```csharp
 ref struct G<T>

--- a/src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs
@@ -3275,7 +3275,7 @@ namespace Microsoft.CodeAnalysis.CSharp
 
             Conversion conversion = Conversions.ClassifyBuiltInConversion(operandType, targetType, isChecked: CheckOverflowAtRuntime, ref useSiteInfo);
             diagnostics.Add(node, useSiteInfo);
-            ReportIsOperatorConstantWarnings(node, diagnostics, operandType, targetType, conversion.Kind, operand.ConstantValue);
+            ReportIsOperatorDiagnostics(node, diagnostics, operandType, targetType, conversion.Kind, operand.ConstantValue);
             return new BoundIsOperator(node, operand, typeExpression, conversion.Kind, resultType);
 
             bool tryBindAsType(
@@ -3293,7 +3293,7 @@ namespace Microsoft.CodeAnalysis.CSharp
 
         }
 
-        private static void ReportIsOperatorConstantWarnings(
+        private static void ReportIsOperatorDiagnostics(
             CSharpSyntaxNode syntax,
             BindingDiagnosticBag diagnostics,
             TypeSymbol operandType,
@@ -3309,13 +3309,27 @@ namespace Microsoft.CodeAnalysis.CSharp
             ConstantValue constantValue = GetIsOperatorConstantResult(operandType, targetType, conversionKind, operandConstantValue);
             if (constantValue != null)
             {
-                Debug.Assert(constantValue == ConstantValue.True || constantValue == ConstantValue.False);
+                if (constantValue.IsBad)
+                {
+                    Error(diagnostics, ErrorCode.ERR_BadBinaryOps, syntax, "is", operandType, targetType);
+                }
+                else
+                {
+                    Debug.Assert(constantValue == ConstantValue.True || constantValue == ConstantValue.False);
 
-                ErrorCode errorCode = constantValue == ConstantValue.True ? ErrorCode.WRN_IsAlwaysTrue : ErrorCode.WRN_IsAlwaysFalse;
-                Error(diagnostics, errorCode, syntax, targetType);
+                    ErrorCode errorCode = constantValue == ConstantValue.True ? ErrorCode.WRN_IsAlwaysTrue : ErrorCode.WRN_IsAlwaysFalse;
+                    Error(diagnostics, errorCode, syntax, targetType);
+                }
             }
         }
 
+        /// <summary>
+        /// Possible return values:
+        ///  - <see cref="ConstantValue.False"/>
+        ///  - <see cref="ConstantValue.True"/>
+        ///  - <see cref="ConstantValue.Bad"/> - compiler doesn't support the type check, i.e. cannot perform it, even at runtime
+        ///  - 'null' value - result is not known at compile time    
+        /// </summary>
         internal static ConstantValue GetIsOperatorConstantResult(
             TypeSymbol operandType,
             TypeSymbol targetType,
@@ -3445,10 +3459,11 @@ namespace Microsoft.CodeAnalysis.CSharp
                         return ConstantValue.False;
                     }
 
-                    // * Otherwise, if the other type is a restricted type, we know no conversion is possible.
+                    // * If either type is a restricted type, the type check isn't supported because
+                    //   a restricted type cannot be boxed or unboxed into.
                     if (targetType.IsRestrictedType() || operandType.IsRestrictedType())
                     {
-                        return ConstantValue.False;
+                        return ConstantValue.Bad;
                     }
 
                     // * Otherwise, we give up. Though there are other situations in which we can deduce that
@@ -3802,13 +3817,13 @@ namespace Microsoft.CodeAnalysis.CSharp
 
             if (!hasErrors)
             {
-                ReportAsOperatorConstantWarnings(node, diagnostics, operandType, targetType, conversionKind, operandConstantValue);
+                ReportAsOperatorDiagnostics(node, diagnostics, operandType, targetType, conversionKind, operandConstantValue);
             }
 
             return hasErrors;
         }
 
-        private static void ReportAsOperatorConstantWarnings(
+        private static void ReportAsOperatorDiagnostics(
             CSharpSyntaxNode node,
             BindingDiagnosticBag diagnostics,
             TypeSymbol operandType,
@@ -3824,11 +3839,24 @@ namespace Microsoft.CodeAnalysis.CSharp
             ConstantValue constantValue = GetAsOperatorConstantResult(operandType, targetType, conversionKind, operandConstantValue);
             if (constantValue != null)
             {
-                Debug.Assert(constantValue.IsNull);
-                Error(diagnostics, ErrorCode.WRN_AlwaysNull, node, targetType);
+                if (constantValue.IsBad)
+                {
+                    Error(diagnostics, ErrorCode.ERR_BadBinaryOps, node, "as", operandType, targetType);
+                }
+                else
+                {
+                    Debug.Assert(constantValue.IsNull);
+                    Error(diagnostics, ErrorCode.WRN_AlwaysNull, node, targetType);
+                }
             }
         }
 
+        /// <summary>
+        /// Possible return values:
+        ///  - <see cref="ConstantValue.Null"/>
+        ///  - <see cref="ConstantValue.Bad"/> - compiler doesn't support the type check, i.e. cannot perform it, even at runtime
+        ///  - 'null' value - result is not known at compile time    
+        /// </summary>
         internal static ConstantValue GetAsOperatorConstantResult(TypeSymbol operandType, TypeSymbol targetType, ConversionKind conversionKind, ConstantValue operandConstantValue)
         {
             // NOTE:    Even though BoundIsOperator and BoundAsOperator will always have no ConstantValue
@@ -3837,9 +3865,17 @@ namespace Microsoft.CodeAnalysis.CSharp
             // NOTE:    and during rewriting for optimized codegen.
 
             ConstantValue isOperatorConstantResult = GetIsOperatorConstantResult(operandType, targetType, conversionKind, operandConstantValue);
-            if (isOperatorConstantResult != null && !isOperatorConstantResult.BooleanValue)
+            if (isOperatorConstantResult != null)
             {
-                return ConstantValue.Null;
+                if (isOperatorConstantResult.IsBad)
+                {
+                    return isOperatorConstantResult;
+                }
+
+                if (!isOperatorConstantResult.BooleanValue)
+                {
+                    return ConstantValue.Null;
+                }
             }
 
             return null;

--- a/src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs
@@ -595,7 +595,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                     else
                     {
                         RoslynDebug.Assert(expression.Type is { });
-                        if (ExpressionOfTypeMatchesPatternType(Conversions, inputType, expression.Type, ref useSiteInfo, out _, operandConstantValue: null) == false)
+                        ConstantValue match = ExpressionOfTypeMatchesPatternType(Conversions, inputType, expression.Type, ref useSiteInfo, out _, operandConstantValue: null);
+                        if (match == ConstantValue.False || match == ConstantValue.Bad)
                         {
                             diagnostics.Add(ErrorCode.ERR_PatternWrongType, expression.Syntax.Location, inputType, expression.Display);
                             hasErrors = true;
@@ -739,10 +740,10 @@ namespace Microsoft.CodeAnalysis.CSharp
                 }
 
                 CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
-                bool? matchPossible = ExpressionOfTypeMatchesPatternType(
+                ConstantValue matchPossible = ExpressionOfTypeMatchesPatternType(
                     Conversions, inputType, patternType, ref useSiteInfo, out Conversion conversion, operandConstantValue: null, operandCouldBeNull: true);
                 diagnostics.Add(typeSyntax, useSiteInfo);
-                if (matchPossible != false)
+                if (matchPossible != ConstantValue.False && matchPossible != ConstantValue.Bad)
                 {
                     if (!conversion.Exists && (inputType.ContainsTypeParameter() || patternType.ContainsTypeParameter()))
                     {
@@ -771,10 +772,12 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// <summary>
         /// Does an expression of type <paramref name="expressionType"/> "match" a pattern that looks for
         /// type <paramref name="patternType"/>?
-        /// 'true' if the matched type catches all of them, 'false' if it catches none of them, and
-        /// 'null' if it might catch some of them.
+        ///  - <see cref="ConstantValue.True"/> if the matched type catches all of them
+        ///  - <see cref="ConstantValue.False"/> if it catches none of them
+        ///  - <see cref="ConstantValue.Bad"/> - compiler doesn't support the type check, i.e. cannot perform it, even at runtime
+        ///  - 'null' if it might catch some of them.
         /// </summary>
-        internal static bool? ExpressionOfTypeMatchesPatternType(
+        internal static ConstantValue ExpressionOfTypeMatchesPatternType(
             Conversions conversions,
             TypeSymbol expressionType,
             TypeSymbol patternType,
@@ -790,7 +793,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             if (expressionType.Equals(patternType, TypeCompareKind.AllIgnoreOptions))
             {
                 conversion = Conversion.Identity;
-                return true;
+                return ConstantValue.True;
             }
 
             if (expressionType.IsDynamic())
@@ -803,13 +806,9 @@ namespace Microsoft.CodeAnalysis.CSharp
             ConstantValue result = Binder.GetIsOperatorConstantResult(expressionType, patternType, conversion.Kind, operandConstantValue, operandCouldBeNull);
 
             // Don't need to worry about checked user-defined operators
-            Debug.Assert(!conversion.IsUserDefined || result == ConstantValue.False);
+            Debug.Assert(!conversion.IsUserDefined || result == ConstantValue.False || result == ConstantValue.Bad);
 
-            return
-                (result == null) ? (bool?)null :
-                (result == ConstantValue.True) ? true :
-                (result == ConstantValue.False) ? false :
-                throw ExceptionUtilities.UnexpectedValue(result);
+            return result;
         }
 
         private BoundPattern BindDeclarationPattern(

--- a/src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs
+++ b/src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs
@@ -1249,14 +1249,14 @@ namespace Microsoft.CodeAnalysis.CSharp
                         case BoundDagTypeTest t2:
                             {
                                 var useSiteInfo = new CompoundUseSiteInfo<AssemblySymbol>(_diagnostics, _compilation.Assembly);
-                                bool? matches = ExpressionOfTypeMatchesPatternTypeForLearningFromSuccessfulTypeTest(t1.Type, t2.Type, ref useSiteInfo);
-                                if (matches == false)
+                                ConstantValue matches = ExpressionOfTypeMatchesPatternTypeForLearningFromSuccessfulTypeTest(t1.Type, t2.Type, ref useSiteInfo);
+                                if (matches == ConstantValue.False)
                                 {
                                     // If T1 could never be T2
                                     // v is T1 --> !(v is T2)
                                     trueTestPermitsTrueOther = false;
                                 }
-                                else if (matches == true)
+                                else if (matches == ConstantValue.True)
                                 {
                                     // If T1: T2
                                     // v is T1 --> v is T2
@@ -1266,7 +1266,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                                 // If every T2 is a T1, then failure of T1 implies failure of T2.
                                 matches = Binder.ExpressionOfTypeMatchesPatternType(_conversions, t2.Type, t1.Type, ref useSiteInfo, out _);
                                 _diagnostics.Add(syntax, useSiteInfo);
-                                if (matches == true)
+                                if (matches == ConstantValue.True)
                                 {
                                     // If T2: T1
                                     // !(v is T1) --> !(v is T2)
@@ -1487,12 +1487,12 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// of a switch (on the one hand) and a series of if-then-else statements (on the other).
         /// See, for example, https://github.com/dotnet/roslyn/issues/35661
         /// </summary>
-        private bool? ExpressionOfTypeMatchesPatternTypeForLearningFromSuccessfulTypeTest(
+        private ConstantValue ExpressionOfTypeMatchesPatternTypeForLearningFromSuccessfulTypeTest(
             TypeSymbol expressionType,
             TypeSymbol patternType,
             ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
         {
-            bool? result = Binder.ExpressionOfTypeMatchesPatternType(_conversions, expressionType, patternType, ref useSiteInfo, out Conversion conversion);
+            ConstantValue result = Binder.ExpressionOfTypeMatchesPatternType(_conversions, expressionType, patternType, ref useSiteInfo, out Conversion conversion);
             return (!conversion.Exists && isRuntimeSimilar(expressionType, patternType))
                 ? null // runtime and compile-time test behavior differ. Pretend we don't know what happens.
                 : result;

--- a/src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs
+++ b/src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs
@@ -1249,7 +1249,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                         case BoundDagTypeTest t2:
                             {
                                 var useSiteInfo = new CompoundUseSiteInfo<AssemblySymbol>(_diagnostics, _compilation.Assembly);
-                                ConstantValue matches = ExpressionOfTypeMatchesPatternTypeForLearningFromSuccessfulTypeTest(t1.Type, t2.Type, ref useSiteInfo);
+                                ConstantValue? matches = ExpressionOfTypeMatchesPatternTypeForLearningFromSuccessfulTypeTest(t1.Type, t2.Type, ref useSiteInfo);
                                 if (matches == ConstantValue.False)
                                 {
                                     // If T1 could never be T2
@@ -1487,7 +1487,7 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// of a switch (on the one hand) and a series of if-then-else statements (on the other).
         /// See, for example, https://github.com/dotnet/roslyn/issues/35661
         /// </summary>
-        private ConstantValue ExpressionOfTypeMatchesPatternTypeForLearningFromSuccessfulTypeTest(
+        private ConstantValue? ExpressionOfTypeMatchesPatternTypeForLearningFromSuccessfulTypeTest(
             TypeSymbol expressionType,
             TypeSymbol patternType,
             ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)

--- a/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AsOperator.cs
+++ b/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AsOperator.cs
@@ -7,6 +7,7 @@ using System.Diagnostics;
 using Microsoft.CodeAnalysis.CSharp.Symbols;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
@@ -51,6 +52,11 @@ namespace Microsoft.CodeAnalysis.CSharp
 
                 if (constantValue != null)
                 {
+                    if (constantValue.IsBad)
+                    {
+                        throw ExceptionUtilities.UnexpectedValue(constantValue);
+                    }
+
                     Debug.Assert(constantValue.IsNull);
                     BoundExpression result = rewrittenType.IsNullableType() ? new BoundDefaultExpression(syntax, rewrittenType) : MakeLiteral(syntax, constantValue, rewrittenType);
 

--- a/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IsOperator.cs
+++ b/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IsOperator.cs
@@ -7,6 +7,7 @@ using System.Diagnostics;
 using Microsoft.CodeAnalysis.CSharp.Symbols;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
@@ -58,6 +59,11 @@ namespace Microsoft.CodeAnalysis.CSharp
 
                 if (constantValue != null)
                 {
+                    if (constantValue.IsBad)
+                    {
+                        throw ExceptionUtilities.UnexpectedValue(constantValue);
+                    }
+
                     return RewriteConstantIsOperator(syntax, rewrittenOperand, constantValue, rewrittenType);
                 }
                 else if (conversionKind.IsImplicitConversion())

--- a/src/Compilers/CSharp/Test/Emit2/Semantics/PatternMatchingTests.cs
+++ b/src/Compilers/CSharp/Test/Emit2/Semantics/PatternMatchingTests.cs
@@ -11689,5 +11689,91 @@ static class C
 }
 ");
         }
+
+        [Fact]
+        [WorkItem(63085, "https://github.com/dotnet/roslyn/issues/63085")]
+        public void RefStructTypeTest_01()
+        {
+            CreateCompilation(@"
+using System;
+
+new G<int>().Test();
+new G<object>().Test();
+new G<int>().TestPattern();
+new G<object>().TestPattern();
+
+ref struct G<T>
+{
+    public void Test()
+    {
+        if (this is G<int>)
+        {
+            Console.WriteLine(""int"");
+        }
+        else if (this is G<object>)
+        {
+            Console.WriteLine(""object"");
+        }
+        else
+        {
+            Console.WriteLine(""unknown"");
+            Console.WriteLine(typeof(T));
+        }
+    }
+
+    public void TestPattern()
+    {
+        var genericTypePattern = this switch
+        {
+            G<int> => ""int"",
+            G<object> => ""object"",
+            _ => ""unknown""
+        };
+
+        Console.WriteLine(genericTypePattern);
+    }
+}
+").VerifyDiagnostics(
+                // (13,13): error CS0019: Operator 'is' cannot be applied to operands of type 'G<T>' and 'G<int>'
+                //         if (this is G<int>)
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "this is G<int>").WithArguments("is", "G<T>", "G<int>").WithLocation(13, 13),
+                // (17,18): error CS0019: Operator 'is' cannot be applied to operands of type 'G<T>' and 'G<object>'
+                //         else if (this is G<object>)
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "this is G<object>").WithArguments("is", "G<T>", "G<object>").WithLocation(17, 18),
+                // (32,13): error CS8121: An expression of type 'G<T>' cannot be handled by a pattern of type 'G<int>'.
+                //             G<int> => "int",
+                Diagnostic(ErrorCode.ERR_PatternWrongType, "G<int>").WithArguments("G<T>", "G<int>").WithLocation(32, 13),
+                // (33,13): error CS8121: An expression of type 'G<T>' cannot be handled by a pattern of type 'G<object>'.
+                //             G<object> => "object",
+                Diagnostic(ErrorCode.ERR_PatternWrongType, "G<object>").WithArguments("G<T>", "G<object>").WithLocation(33, 13)
+                );
+        }
+
+        [Fact]
+        [WorkItem(63085, "https://github.com/dotnet/roslyn/issues/63085")]
+        public void RefStructTypeTest_02()
+        {
+            CreateCompilation(@"
+ref struct G<T> where T : class
+{
+    public void Test1(T x1)
+    {
+        var y1 = x1 as G<object>;
+    }
+
+    public void Test2(G<object> x2)
+    {
+        var y2 = x2 as T;
+    }
+}
+").VerifyDiagnostics(
+                // (6,18): error CS0077: The as operator must be used with a reference type or nullable type ('G<object>' is a non-nullable value type)
+                //         var y1 = x1 as G<object>;
+                Diagnostic(ErrorCode.ERR_AsMustHaveReferenceType, "x1 as G<object>").WithArguments("G<object>").WithLocation(6, 18),
+                // (11,18): error CS0019: Operator 'as' cannot be applied to operands of type 'G<object>' and 'T'
+                //         var y2 = x2 as T;
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "x2 as T").WithArguments("as", "G<object>", "T").WithLocation(11, 18)
+                );
+        }
     }
 }

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs
@@ -264,9 +264,9 @@ class C
     }
 }");
             comp.VerifyDiagnostics(
-                // (7,13): warning CS0184: The given expression is never of the provided ('Span<int>') type
+                // (7,13): error CS0019: Operator 'is' cannot be applied to operands of type 'T' and 'Span<int>'
                 //         if (t is Span<int>)
-                Diagnostic(ErrorCode.WRN_IsAlwaysFalse, "t is Span<int>").WithArguments("System.Span<int>").WithLocation(7, 13),
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "t is Span<int>").WithArguments("is", "T", "System.Span<int>").WithLocation(7, 13),
                 // (9,18): error CS8121: An expression of type 'T' cannot be handled by a pattern of type 'Span<int>'.
                 //         if (t is Span<int> s)
                 Diagnostic(ErrorCode.ERR_PatternWrongType, "Span<int>").WithArguments("T", "System.Span<int>").WithLocation(9, 18));
@@ -308,9 +308,9 @@ class C
     }
 }");
             comp.VerifyDiagnostics(
-                // (7,13): warning CS0184: The given expression is never of the provided ('T') type
+                // (7,13): error CS0019: Operator 'is' cannot be applied to operands of type 'Span<int>' and 'T'
                 //         if (s is T) { }
-                Diagnostic(ErrorCode.WRN_IsAlwaysFalse, "s is T").WithArguments("T").WithLocation(7, 13),
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "s is T").WithArguments("is", "System.Span<int>", "T").WithLocation(7, 13),
                 // (8,18): error CS8121: An expression of type 'Span<int>' cannot be handled by a pattern of type 'T'.
                 //         if (s is T t) { }
                 Diagnostic(ErrorCode.ERR_PatternWrongType, "T").WithArguments("System.Span<int>", "T").WithLocation(8, 18));
@@ -330,9 +330,9 @@ class C
     }
 }");
             comp.VerifyDiagnostics(
-                // (7,13): warning CS0184: The given expression is never of the provided ('T') type
+                // (7,13): error CS0019: Operator 'is' cannot be applied to operands of type 'Span<int>' and 'T'
                 //         if (s is T) { }
-                Diagnostic(ErrorCode.WRN_IsAlwaysFalse, "s is T").WithArguments("T").WithLocation(7, 13),
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "s is T").WithArguments("is", "System.Span<int>", "T").WithLocation(7, 13),
                 // (8,18): error CS8121: An expression of type 'Span<int>' cannot be handled by a pattern of type 'T'.
                 //         if (s is T t) { }
                 Diagnostic(ErrorCode.ERR_PatternWrongType, "T").WithArguments("System.Span<int>", "T").WithLocation(8, 18));

